crates/farming/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+//! # Farming Module
+//! Root can create reward schedules which payout incentives
+//! on a per period basis. Users can stake LP tokens, such as
+//! those generated from an AMM or lending protocol to receive
+//! these rewards by claiming.
+
+// #![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod default_weights;
+pub use default_weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{dispatch::DispatchResult, traits::Get, transactional, weights::Weight, PalletId, RuntimeDebug};
+use orml_traits::{MultiCurrency, MultiReservableCurrency};
+use primitives::CurrencyId;
+use reward::RewardsApi;
+use scale_info::TypeInfo;
+use sp_runtime::{
+    traits::{AccountIdConversion, AtLeast32Bit, Saturating, Zero},
+    ArithmeticError,
+};
+
+pub use pallet::*;
+
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+pub type CurrencyIdOf<T> =
+    <<T as Config>::MultiCurrency as MultiCurrency<<T as frame_system::Config>::AccountId>>::CurrencyId;
+
+type BalanceOf<T> = <<T as Config>::MultiCurrency as MultiCurrency<AccountIdOf<T>>>::Balance;
+
+pub(crate) type RewardScheduleOf<T> = RewardSchedule<<T as frame_system::Config>::BlockNumber, BalanceOf<T>>;
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct RewardSchedule<BlockNumber, Balance: MaxEncodedLen> {
+    /// Block height to start after
+    pub start_height: BlockNumber,
+    /// Number of blocks between distribution
+    pub period: BlockNumber,
+    /// Number of periods remaining
+    pub period_count: u32,
+    /// Amount of tokens to release
+    #[codec(compact)]
+    pub per_period: Balance,
+}
+
+impl<BlockNumber: AtLeast32Bit + Copy, Balance: AtLeast32Bit + MaxEncodedLen + Copy>
+    RewardSchedule<BlockNumber, Balance>
+{
+    /// Returns total amount to distribute, `None` if calculation overflows
+    pub fn total(&self) -> Option<Balance> {
```

Comment:
Because there is no `Config` bound on this type or impl, see [`VestingSchedule`](https://github.com/open-web3-stack/open-runtime-module-library/blob/04e67e6a74510a1f633a41bb98a65299b2921231/vesting/src/lib.rs#L86) where I took inspiration

Reply:
I think you _could_ add one, but it might be a bit clunky so nvm